feature/21-fuzzy-matching-with-tollerance (#23)
exasol/hamcrest-resultset-matcher: doc/user_guide/user_guide.md

# Hamcrest ResultSet Matcher User Guide

> The `ResultSet` matcher is the implementation of a [Java Hamcrest matcher](JavaHamcrest). The Hamcrest (an anagram of "

Author: why not it's easier to read for everyone and in the markdown viewers (like GitHub) it's the same.
The long lines only make it hard to read.

> The `ResultSet` matcher is the implementation of a [Java Hamcrest matcher](JavaHamcrest). The Hamcrest (an anagram of "

Author: Oh. I just realised some formater broke the lines in the sentence... That's of course not nice

> The `ResultSet` matcher is the implementation of a [Java Hamcrest matcher](JavaHamcrest). The Hamcrest (an anagram of "

Reviewer: 1. All Markdown readers I came across support soft-wrapping.
1. If you add new text, multiple lines shift, leading to a broader than necessary diff.
1. It's easier to see real paragraph changes when line breaks are reserved for them.

    void testTableContents() {
        // ...
        final ResulSet result = statement.execute("SELECT * FROM CUSTOMERS");

        assertThat(result, table()
                .row(1, "JOHN", "DOE")
                .row(2, "JANE", "SMITH)

>                         .matches());

Reviewer: Why the extra indentation?
It's still the same level.

```java
assertThat(result,
        table()
        .row("fred",matchesPattern("[0-9A-F]+"),greaterThanOrEqualTo(0))
        .matches();
```

> Of course you can nest matcher in the nested matcher. That's the beauty of hamcrest.

Reviewer: ```suggestion
Of course you can nest matchers in the nested matcher. That's the beauty of Hamcrest.
```

### Matching Floating-Point Numbers

> When comparing floating-point numbers you sometimes want only to compare with a given precision. So you don't want to check that all decimal digits are exactly equal but match with a certain tolerance.

Reviewer: ```suggestion
When comparing floating-point numbers you might want to compare the actual value against the expectation within a given tolerance. That way actual and expected don't need to be an exact match &mdash; it just needs to be close enough.
```

Wou can do that by:

```java

> assertThat(result,
>         table()
>         table().row(
>         CellMatcherFactory.cellMatcher(1.234,TypeMatchMode.STRICT,BigDecimal.valueOf(0.001)))
>         .matches();
> ```

Reviewer: ```suggestion
final BigDecimal tolerance = BigDecimal.valueOf(0.001);
assertThat(result,
        table()
        .row(CellMatcherFactory.cellMatcher(1.234, TypeMatchMode.STRICT, tolerance))
        .matches();
```
